### PR TITLE
completed Scrabble project

### PR DESCRIPTION
**Pull request by a-lmx**
This is for the dinner level.

Run rspec in the project root folder for tests. Run lib/scrabble_module.rb in IRB/PRY to play with the code.

In this project, I worked on using appropriate enumerable methods (rather than .each loops every time), so if you see any that are not as appropriate as something else, I'd appreciate that feedback.

I also did some method refactoring after I got initial versions to work (on Scrabble.score(word) and Scrabble.highest_word_from(array_of_words)), to pull out individual pieces of logic into their own methods. Is this overkill, or a good application of the single responsibility principle?

Thanks in advance.

### CODE REVIEW

File: lib/scrabble/scrabble.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	  		# guard clause for empty string or nil input
+		  	return false if word == "" || word == nil
+
+		  	word_letters = self.process_input(word)
+		  	return false if word_letters.find { |letter| !VALID_LETTERS.include?(letter) }
+			
+		  	true
+	  	end
+
+	  	def self.check_length(word)
```
Comment (dougcole) [inline lib/scrabble/scrabble.rb]: I would consider a more descriptive name. What does `check` mean in this context. Also, ruby allows methods to end with a `?`, which is a great addition for methods that return boolean values like this! Maybe `valid_word length?`? I don't love that name either, but I think it's better.

Comment (dougcole) [inline lib/scrabble/scrabble.rb]: Actually looking below at how this is used, what if you raised an exception inside this method? You could raise an `ArgumentError` if the word is too long and rename the method something like `validate_length`.

File: lib/scrabble/scrabble.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+		  	return false unless word.length <= 7
+		  	true
```
Comment (dougcole) [inline lib/scrabble/scrabble.rb]: Ruby methods returns the value of the last expression, so rather than have these two lines you could just have `word.length <= 7` which will have the same return value.

File: lib/scrabble/scrabble.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	  	end
+
+		def self.get_points(word)
+		  	word_letters = Scrabble.process_input(word)
+
+		  	word_score = word_letters.inject(0) do |sum, letter|
```
Comment (dougcole) [inline lib/scrabble/scrabble.rb]: nice use of inject! :+1:

File: lib/scrabble/scrabble.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		  	"y" => 4, "z" => 10
+	  	}
+
+	  	VALID_LETTERS = POINTS.keys.reduce(:+)
+
+	  	# def initialize
+	  	# end
+
+	 	def self.process_input(word)
+	 		word_letters = word.downcase.split("")
```
Comment (dougcole) [inline lib/scrabble/scrabble.rb]: since this variable is never used, you can skip declaring it.

File: lib/scrabble/scrabble.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+	 	POINTS = {
+		  	"a" => 1, "b" => 3, "c" => 3, "d" => 2, "e" => 1, "f" => 4,
+		  	"g" => 2, "h" => 4, "i" => 1, "j" => 8, "k" => 5, "l" => 1,
+		  	"m" => 3, "n" => 1, "o" => 1, "p" => 3, "q" => 10, "r" => 1,
+		  	"s" => 1, "t" => 1, "u" => 1, "v" => 4, "w" => 4, "x" => 8,
+		  	"y" => 4, "z" => 10
+	  	}
+
+	  	VALID_LETTERS = POINTS.keys.reduce(:+)
```
Comment (dougcole) [inline lib/scrabble/scrabble.rb]: This works, but a more idiomatic way of doing the same thing would be to use `join` instead of `reduce(:+)`

File: lib/scrabble/scrabble.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		  		sum + POINTS[letter]
+		  	end
+		end
+
+	  	def self.score(word)
+		  	# guard clause for non-alphabetic user input
+	  		return "ERROR -- invalid input" unless Scrabble.valid_input?(word)
+
+	  		# guard clause for word over 7 letters
+	  		return "ERROR -- too long" unless Scrabble.check_length(word)
```
Comment (dougcole) [inline lib/scrabble/scrabble.rb]: It's worth thinking about who these errors are intended for. If they are for someone who is playing the game, I'd make them a little more verbose. If they are intended for other users of this class, I'd consider raising an exception and letting them deal with it.

File: spec/scrabble_spec.rb [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+		it "returns #{error} when user input is #{word}" do
+			expect(Scrabble::Scrabble.score(word)).to eq(error)
+		end
+	end
+
+	it "rejects words over 7 letters long" do
+		expect(Scrabble::Scrabble.score("elephant")).to eq("ERROR -- too long")
+	end
+
+end
```
Comment (dougcole) [inline spec/scrabble_spec.rb:79]: These specs look great!

File: lib/scrabble/scrabble.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+	  		# return score for valid word
+	  		Scrabble.get_points(word)
+	  	end
+
+	  	def self.get_high_scores(array_of_words)
+	  		score_groups = array_of_words.group_by do |word|
+		  		return nil unless Scrabble.valid_input?(word)
```
Comment (dougcole) [inline lib/scrabble/scrabble.rb]: If this raised an exception, you could skip the additional check of the return value down on line 81.

File: lib/scrabble/scrabble.rb [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+		  		return nil unless Scrabble.valid_input?(word)	  			
+	  			Scrabble.score(word)
+	  		end
+
+	  		top_score = score_groups.keys.max
+
+	  		score_groups[top_score].flatten	# top words
+	  	end
+
+	  	def self.seven_letter_word(high_score_words)
```
Comment (dougcole) [inline lib/scrabble/scrabble.rb]: nitpick, but I'd rename the argument of this method as it's more general than just a list of high score words.

File: lib/scrabble/scrabble.rb [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+	  		seven_letter_word = high_score_words.find do 
+		  		|word, score| word.length == 7 
+		  	end
+	  	end
+
+	  	def self.highest_word(high_score_words)
```
Comment (dougcole) [inline lib/scrabble/scrabble.rb]: I feel like this name is a little misleading as it doesn't really describe what the method does.

**Comment by dougcole** [comment]
gotta run - I'll add more feedback later!

File: lib/scrabble/player.rb [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+			total_score
+		end
+
+		def won?
+			return true if total_score >= 100
+			false
+		end
+
+		def highest_scoring_word
+			return false if @plays == []
```
Comment (dougcole) [inline lib/scrabble/player.rb]: you could also use `@plays.empty?` which reads a little better, imo.

File: lib/scrabble/player.rb [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+			@name 	= name
+			@plays 	= []
+			@tiles 	= []
+		end
+
+		def play(word)
+			# guard clause for if word is invalid
+			return "ERROR -- invalid input" unless Scrabble.valid_input?(word)
+			# guard clause for if player has already won
+			return false if won? == true
```
Comment (dougcole) [inline lib/scrabble/player.rb]: `won?` is "truthy" already, so you don't need to compare it to `true`. `if won?` will give the same result.

**Comment by dougcole** [comment]
With regards to refactoring to smaller methods being overkill vs good application of SRP... that's always going to be a judgement call. Smaller methods make for easier to make changes in the future and reuse pieces, but can make it harder to follow the overall logic of a program.

In my opinion `highest_score_from` was a little hard to follow, but I'm sure other people would disagree with me. Other than that I thought they all looked great! I added a lot of small suggestions, but overall the code was concise, easy to follow and had great specs. Nice work!

Let me know if you have any questions.

File: lib/scrabble/scrabble.rb (hunk 1 above)
Comment (a-lmx) [inline lib/scrabble/scrabble.rb]: I like this. I'll fix the code & test and push the change.

File: lib/scrabble/scrabble.rb (hunk 2 above)
Comment (a-lmx) [inline lib/scrabble/scrabble.rb]: Mind. BLOWN.

File: lib/scrabble/scrabble.rb (hunk 4 above)
Comment (a-lmx) [inline lib/scrabble/scrabble.rb]: I'm still getting used to that, will pay more attention going forward. I think I can also use .chars instead of split(""), no?

File: lib/scrabble/scrabble.rb (hunk 5 above)
Comment (a-lmx) [inline lib/scrabble/scrabble.rb]: Got it, that's much clearer.

File: lib/scrabble/scrabble.rb (hunk 6 above)
Comment (a-lmx) [inline lib/scrabble/scrabble.rb]: I'll decide they're for other users of the class, and make the validation errors argument errors instead.

File: lib/scrabble/scrabble.rb (hunk 10 above)
Comment (a-lmx) [inline lib/scrabble/scrabble.rb]: Do you mean .highest_word, or .seven_letter_word ? I think they're both clumsy, just curious which you were referring to.

File: lib/scrabble/scrabble.rb (hunk 9 above)
Comment (a-lmx) [inline lib/scrabble/scrabble.rb]: Good point. The method could be run on any group of words.

File: lib/scrabble/player.rb (hunk 11 above)
Comment (a-lmx) [inline lib/scrabble/player.rb]: Thanks, I hadn't known that one before.

File: lib/scrabble/player.rb (hunk 12 above)
Comment (a-lmx) [inline lib/scrabble/player.rb]: So compact -- I love it.

**Comment by a-lmx** [comment]
Thanks so much for your comments, Doug. They were really helpful, not just for learning some new techniques, but also for thinking about the methods independently and not just how they contribute to this particular project. I refactored highest_score_from to have clearer flow, hope it reads better now.

File: lib/scrabble/scrabble.rb (hunk 10 above)
Comment (dougcole) [inline lib/scrabble/scrabble.rb]: I meant `highest_word`, it's not clear what highest means in this context.

Comment (a-lmx) [inline lib/scrabble/scrabble.rb]: Got it. I renamed it `.get_shortest_word` since that's what it does.
